model/riscv_insts_zfa.sail

```diff
@@ -685,54 +685,57 @@ function clause execute(RISCV_FLTQ_D(rs2, rs1, rd)) = {
+                                  else integer;
+
+    /* raise FP exception flags, honoring the precedence of nV > nX */
+    let flags : bits(5) = if (true_exp > 31)         then nvFlag()
+                     else if (fractional != zeros()) then nxFlag()
+                     else                            zeros();
+
+    (flags, to_bits(32, result))
+}
+
 /*
  * Implement float64 to int32 conversion without saturation, the
  * result is supplied modulo 2^32.
- * Rounding mode for FCVTMOD.W.D is always RTZ(Round toward zero).
  */
-val      fcvtmod_helper : forall 'm, 'm == 64. (bits('m)) -> (bits(5), xlenbits)
+val      fcvtmod_helper : forall 'm, 'm == 64. (bits('m)) -> (bits(5), bits(32))
 function fcvtmod_helper(x64) = {
   let (sign, exp, mant) = fsplit_D(x64);
-  fflags		= fcsr.FFLAGS();
-  result : bits(64)	= undefined;
-
-  /* Handle special cases. */
-  if (exp == zeros())
-  then (nvFlag(), zeros())
-  /* NaN or Infinity. */
-  else if (exp == ones())
-  then (nvFlag(), zeros())
-  else {
-    let true_exp : int = unsigned(sub_vec_int(exp, 1023));
-    let shift    : int = true_exp - 52;
-    result	       = to_bits(64, unsigned(mant)) | ones() << 52;
-
-    /* Shift the fraction into place and set NX flag. */
-    if (shift >= 64 | shift <= -64) then {
-	/* The fraction is shifted out entirely. */
-	result = zeros();
-	fflags = fflags | nxFlag();
-    } else if (shift >= 0) then {
-	/* Raise NX if bit 52 got shifted out. */
-	result = result  << to_bits(64, shift);
-	if (unsigned(result) >= 12)
-	then fflags = fflags | nxFlag();
-    } else {
-	result = result >> to_bits(64, negate(shift));
-	/* Raise NX if bits got shifted out. */
-	if ((result << to_bits(64, negate(shift))) != result)
-	then fflags = fflags | nxFlag();
-    };
-
-    if (sign == 0b1)
-    then result = negate_D(result);
-
-    let result32b = to_bits(32, unsigned(result));
-    if (to_bits(64, unsigned(result32b)) != result)
-    then fflags = fflags | nvFlag();
-
-    (fflags, EXTS(result[31..0]))
-  };
+
+  /* detect the non-normal cases */
+  let is_subnorm    = (exp == zeros() & mant != zeros());
+  let is_zero       = (exp == zeros() & mant == zeros());
+  let is_nan_or_inf = (exp == ones());
+
+  /* For normal numbers, the mantissa and exponent use special encoding */
+  let true_mant     = 0b1 @ mant;  /* concatenate with the implicit bit */
+  let true_exp      = unsigned(exp) - 1023;   /* undo the offset-binary */
+
+  /* Detect the trivial cases for normal numbers */
+  let is_too_large  = true_exp >= 84;  /* bits will be 'multiplied' out */
+  let is_too_small  = true_exp <  0;   /* bits will be 'divided' out    */
+
+       if (is_zero)       then (zeros(),  zeros())
+  else if (is_subnorm)    then (nxFlag(), zeros())
+  else if (is_nan_or_inf) then (nvFlag(), zeros())
+  else if (is_too_large)  then (nvFlag(), zeros())
+  else if (is_too_small)  then (nxFlag(), zeros())
+  else fcvtmod_helper_inner(sign, true_exp, true_mant)
```

Comment:
Hm, is there a reason you chose to outline this rather than inline it? Not a huge fan of the "_inner" suffix, may benefit from a better description if kept outlined, something about the number being "regular" (deliberately avoiding the word normal there...).

Reply:
Inlined again in the new revision.